internal/servers/controller/handlers/managed_groups/managed_group_service.go

```diff
@@ -0,0 +1,661 @@
+	if err != nil {
+		return nil, errors.Wrap(err, op, errors.WithMsg("unable to create managed group"))
+	}
+	if out == nil {
+		return nil, handlers.ApiErrorWithCodeAndMessage(codes.Internal, "Unable to create managed group but no error returned from repository.")
+	}
+	return out, nil
+}
+
+func (s Service) createInRepo(ctx context.Context, am auth.AuthMethod, item *pb.ManagedGroup) (auth.ManagedGroup, error) {
+	const op = "managed_groups.(Service).createInRepo"
+	if item == nil {
+		return nil, errors.New(errors.InvalidParameter, op, "missing item")
+	}
+	var out auth.ManagedGroup
+	switch auth.SubtypeFromId(am.GetPublicId()) {
+	case auth.OidcSubtype:
+		am, err := s.createOidcInRepo(ctx, am, item)
+		if err != nil {
+			return nil, errors.Wrap(err, op)
+		}
+		if am == nil {
+			return nil, handlers.ApiErrorWithCodeAndMessage(codes.Internal, "Unable to create managed group but no error returned from repository.")
+		}
+		out = am
+	}
+	return out, nil
+}
+
+func (s Service) updateOidcInRepo(ctx context.Context, scopeId, amId, id string, mask []string, item *pb.ManagedGroup) (*oidc.ManagedGroup, error) {
+	const op = "managed_groups.(Service).updateOidcInRepo"
+	if item == nil {
+		return nil, errors.New(errors.InvalidParameter, op, "nil managed group.")
+	}
+	mg := oidc.AllocManagedGroup()
+	mg.PublicId = id
+	if item.GetName() != nil {
+		mg.Name = item.GetName().GetValue()
+	}
+	if item.GetDescription() != nil {
+		mg.Description = item.GetDescription().GetValue()
+	}
+	if apiAttr := item.GetAttributes(); apiAttr != nil {
+		attrs := &pb.OidcManagedGroupAttributes{}
+		if err := handlers.StructToProto(apiAttr, attrs); err != nil {
+			return nil, handlers.InvalidArgumentErrorf("Error in provided request.",
+				map[string]string{attributesField: "Attribute fields do not match the expected format."})
+		}
+		// Set this regardless; it'll only take effect if the masks contain the value
+		mg.Filter = attrs.Filter
+	}
+
+	version := item.GetVersion()
+
+	dbMask := oidcMaskManager.Translate(mask)
+	if len(dbMask) == 0 {
+		return nil, handlers.InvalidArgumentErrorf("No valid fields included in the update mask.", map[string]string{"update_mask": "No valid fields provided in the update mask."})
+	}
+	repo, err := s.oidcRepoFn()
+	if err != nil {
+		return nil, errors.Wrap(err, op)
+	}
+	out, rowsUpdated, err := repo.UpdateManagedGroup(ctx, scopeId, mg, version, dbMask)
+	if err != nil {
+		return nil, errors.Wrap(err, op, errors.WithMsg("unable to update managed group"))
+	}
+	if rowsUpdated == 0 {
+		return nil, handlers.NotFoundErrorf("Managed Group %q doesn't exist or incorrect version provided.", id)
+	}
+	return out, nil
+}
+
+func (s Service) updateInRepo(ctx context.Context, scopeId, authMethodId string, req *pbs.UpdateManagedGroupRequest) (auth.ManagedGroup, error) {
+	const op = "managed_groups.(Service).updateInRepo"
+	var out auth.ManagedGroup
+	switch auth.SubtypeFromId(req.GetId()) {
```

Comment:
ty for using the existing subtype pattern here, even though there's currently only one type of managed group.

Reply:
👍